### DISCUSSION TITLE
YALMIP solver

### DISCUSSION THREAD

**Discussion by ArjunM21** [yalmip/YALMIP#1028] [Optimization modelling] [answered]
Hey all.
So I had previously discussed a question of mine involving YALMIP solvers to solve an optimization problem from a research paper. I had gotten a good answer and proceeded with it. Now, after modifying it as needed and confirming that its working, I have run into  a small problem. Please see the code I run below.

`clc;
clear all;
close all;
C = [1 0 0 ; 1 1 -1; 1 1 -1; 1 1 -1; 0 1 0; 1 0 -1; 0 1 -1; 0 0 1]; %1,5,8
% C = [1 0 0 ; 0 1 0; 0 1 0; 0 1 0; 0 0 1; 0 1 -1;-1 1 0; 1 -1 1]; %1,2,5
sdpvar t;
q = binvar(8,1);
Q = diag(q);
q0 = [1;0;0;0;1;1;0;0];
Q0 = diag(q0);
Model = [sum(q)==3];
ch = 100;

while(ch > 0)
    P0 = inv(((Q0)*C)'*C);
    G0 = trace(C/(C'*Q0*C)*C'); %G(qbar) from the paper
    display(G0);    
    Gdif0 = C*P0*C'*C*P0*C'; %deltaG(q-qbar) from paper
    Model = [Model, G0 + Gdif0*(q-q0) <= t];
    ss = optimize(Model,t);
    q0 = value(q);
    Q0 = diag(q0);
    Qt = sum(q0);
    t0 = value(t);
    display(trace(C/(C'*Q0*C)*C'));
    ch = abs(t0 - trace(C/(C'*Q0*C)*C'));
end `

The best answer that it should ideally converge to is G0 = 11, but now its stuck at 12. The code should ideally take very less time to run (approx. 70 steps maybe?), but now, it seems to not converge anywhere. Could you help me understand why?Also pls help me understand why the loop does not converge? Thank you in advance

**Comment by johanlofberg** [comment, marked as answer]
Is the method guaranteed to find the globally optimal solution? I don't think so judging from the first sentence "Here the two methods to arrive at near optimal solutions are highlighted". Also note that your code finds different solutions depending on the MILP solver (12 with Mosek, 14 with Gurobi). Using q0 = q0(randperm(8)) shows that it sometimes manages to find the global solution, and sometimes not, depending on the initial guess. 

You should round the computed solution to ensure it is binary (most MILP solvers only promises to return a solution close to binary, within specified tolerances). I can easily see a loop like this failing due to missing this

Starting in an optimal solution, and your linearization is feasible and terminates with optimal solution (indicating your code is correct). 

```
Y=sdpvar(8)
optimize([Model, [Y C;C' C'*Q*C]>=0],trace(Y))
q0 = round(value(q))
Q0 = diag(q0);

while...
```

**Reply by ArjunM21** [reply]
Thank you for the reply. But implementing this did not solve the issue. Could you tell me why this loop keeps going forever? Like you said, it did solve it quite easily in certain situations depending on the C matrix or the initial value. But this problem is supposed to solve in very less number of steps.

**Reply by johanlofberg** [reply]
Well, we do not know which solver you are using, and ch can easily be nonzero even though the iterations have converged up to numerical tolerances. Checking for non-zero is a bad idea as you are reasoning with floating-point numbers

**Reply by ArjunM21** [reply]
So what do you think I should do instead? Also, how can I change the solvers as needed?

**Reply by johanlofberg** [reply]
Well, to begin with, is it guaranteed to converge to 0? If not, it makes no sense to check for closeness to 0, but you should check for a lack of change.

If it is guaranteed to converge to 0 (then your code is wrong as it does not, t0 is around 14 while G0 is around 12 in the iterations) you would have to check for closeness to 0, not exactly 0

**Reply by johanlofberg** [reply]
Your code is clearly wrong, as G0 + Gdif0*(q-q0)  is a vector (Gdif0 is a matrix).

**Reply by ArjunM21** [reply]
Hey thank you. I knew something was wrong, but this is  how it says in the paper. The dimensions do not match i guess, but the code runs without error and even gives the right answer. Could you maybe help me code the right way? G is a 1x1 value, as it is trace of a matrix, deltaG is supposed to be a 1x1 value as it is again the trace. q and q0 are both 8x1 vector. But that is exactly how the constraint is written in the paper.

**Reply by johanlofberg** [reply]
Then you have to derive the correct result or contact the authors for errata. MATLAB does not complain as adding a vector and scalar is ok algebraically. Gdif0 has to be a row vector for the math to make sense
